### ISSUE TITLE
Support for Secert Params in Materials

### ISSUE THREAD

**Issue by maheshp** [gocd/gocd#6014]
Part of #5830

- [x] 1. Users should have an ability to define secrets in material url or password field.(PR #5991)

- [x] 2. In a material URL, the secret params can be used only for password e.g https://username:{{SECRET:[secret_config_id][token]}}@foo.bar/baz.git
       If a secret param is used in any other part of the URL it would be a ignored and might lead to an invalid URL. Ideally a check connection should be able to catch these errors.
      - Validation for this is added in PR #6050  

- [x] 3. **Fingerprint, hashcode and equals on Materials** 
        Git, Hg, SVN, TFS Materials allow configuring username and pasword in the url. The configured url is exposed as url_for_commandline and url_for_display(where the password is masked). url_for_commandline is same as the configured url and is used for generating the fingerprint, for equals and hashcode. 
      By adding support for secret params, material url's can be configured to have secret_params. If we want to continue using the url_for_commandline to generate the fingerprint, hashcode and for equals the secret_params need to be resolved before using it. **Instead, use the configured url(unresolved url) to generate the fingerprint, hashcode and for equals.** Therefore going forward, materials will support the following,
  - `configured url` - unresolved url used for fingerprint, hashcode and equals.
  - `url_for_commandline` - resolved url used for material updates.
  - `url_for_display` - used wherever the url is logged/displayed with the password masked.

- [x] 4. Ensure secret_params once resolved are masked when logging, in server health messages, test connection.

- [x] 5. Test connection should work as is for url with secret_params. - (PR #6050)

- [x] 6. All the post commit webhooks should continue to work as is, when materials are configured with secret_params.(PR #5991)

- [x] 7. Secret param resolution
        a. Material url's should be resolved before MDU - failure of resolution should be logged to go-server.log as well as should show up in server health message.
        b. Material url's should be resolved before assigning a job to an agent. Failure of resolution should fail job assignment and an appropriate message should be logged in the job console.
        c. Material url's should be resolved before an MDU on manual trigger of pipeline. Failure should be logged to server health message.
        d. Material url's should be resolved before test connection.
        e. Material url's should be resolved before MDU on notification to material webhook endpoints.
        f. SVN material expansion should work as is if material URL has secrets 
        g. Post-commit hook API should work as is when the material is configured with secret params

- [x] 8. Validation(PR #6050)
  - [x]  a. Ensure secrets params are provided only for Password in a material URL.
  - [x]  b. Ensure the secret param has a valid secret_config_id

- [x] 9. Support for Secret Param defined in config_repos.
- [x] 10. A fingerprint for existing material should not change post GoCD server upgrade.

**Comment by arvindsv** [comment]
> Ensure secrets params are provided only for Username or Password in a material URL.

Password-only. Right?

**Comment by bdpiparva** [comment]
@arvindsv - Currently secret param is supported in username and password both. I will make a change to support it only in a password.

**Comment by adityasood** [comment]
@maheshp @bdpiparva for this

>  Material url's should be resolved before assigning a job to an agent. Failure of resolution should fail job assignment and an appropriate message should be logged in the job console.

When we say should fail job assignment, does this mean this job will not be assigned to an agent until all secret-params are resolved? If yes, I see this is not the case today, job is assigned to an agent irrespective of the secret param resolution. If this is not implemented can we move this out and mark as something to-do?

**Comment by maheshp** [comment]
> When we say should fail job assignment, does this mean this job will not be assigned to an agent until all secret-params are resolved? 

Yes, a given job can have multiple secret params defined in either environment variables or materials. All the secret params should be resolved before we can assign a job to an agent.

You are right bout the jobs being assigned to the agent in spite of the secret params being unresolved. The are 2 reasons for this,
1. The file  based secrets plugin should ideally return an error response if it is unable to resolve a secret param, which it does not now. The plugin should implement https://github.com/gocd/gocd/issues/6018
2. GoCD core did not have a check to verify if all the secrets params are resolved when it makes a secret lookup call, this has been implement as part of https://github.com/gocd/gocd/pull/6099

You should be able to verify this behaviour now.
